### DISCUSSION TITLE
invalid URL: Undefined when requesting realtime subscription

### DISCUSSION THREAD

**Discussion by react-ghost** [supabase/supabase#6522] [Questions] [answered]
When attempting to establish a realtime subscription via a useEffect hook in react-native, I am hit with this message

`TypeError: Invalid URL: undefined`

I don't understand why this would happen here as I am successful in making calls to my database in every form, insert, create, update, delete etc.

I am ONLY hit with this message right now when calling on the realtime database.

I have no useful error message to diagnose the problem, other than the slight amount of traceability offered by the screenshot that is posted below. I can see from this screenshot that the issue is coming from:

`\node_modules\whatwg-url-without-unicode\lib\URL-impl.js (21:13)`

The lines of code in this file are:

```
const parsedURL = usm.basicURLParse(url, { baseURL: parsedBase });
    if (parsedURL === null) {
      throw new TypeError(`Invalid URL: ${url}`);
    }
```
To Reproduce
Here is what my client setup looks like:

```
import { SUPABASE_URL, SUPABASE_API } from "@env";
import { createClient } from "@supabase/supabase-js";
import AsyncStorage from "@react-native-async-storage/async-storage";

const supabase = createClient(SUPABASE_URL, SUPABASE_API, {
  localStorage: AsyncStorage,
  detectSessionInUrl: false,
});

export { supabase };
```

In a react context, I am utilizing a useEffect hook to call this line:

```
useEffect(() => {
    const sub = supabase.from('favorites').on('*', (payload) => console.log(payload)).subscribe();
    return () => supabase.removeSubscription(sub);
  }, [])
```

This context provider wraps the root of my app, which is a method I've used before. I have tried moving around the provider to different parts of the app. At first, I maybe thought this was being called before supabase was initialized but it appears to make no difference where the provider is. I have ensured that replication is on.

What's more weird is that the Supabase web UI is showing me that Realtime connection requests are being made - I can see this in my dashboard that has stats pertaining to API calls, auth request etc. etc.

![Photo Apr 21, 1 22 55 PM](https://user-images.githubusercontent.com/69426533/164540843-d765e349-9bb0-448d-98eb-b8f66c2c46fd.png)

**Comment by GaryAustin1** [comment, marked as answer]
Can you look in your supabase object returned from the client init and see what the base url is?
If there is an extra character like /n I believe Supabase.js still works, but I know storage.js does not.  Realtime.js also may not handle it.
https://github.com/supabase/supabase/issues/5819

**Reply by react-ghost** [reply]
Hey there. Thanks for responding.

I'm not seeing anything out of the ordinary with the URL, no newlines or anything like that.

I see multiple url designations here, all utilizing the same base url here, presumably realtimeUrl from the supabase client object is the one that's getting used? I do not have knowledge of what these various urls are used for. 

It's like wss://projecturl.supabase.co/realtime/v1

**Reply by GaryAustin1** [reply]
That seems OK...  Got nothing else for you.

**Reply by w3b6x9** [reply]
> wss://projecturl.supabase.co/realtime/v1

@adrienclay36 That is the correct url that Realtime uses.

How was `whatwg-url-without-unicode` package introduced into your app? Did you directly install the package or is it used by another dependency?

I don't see anything wrong with your code.

**Reply by react-ghost** [reply]
@w3b6x9 I did not install it manually - so it had to come from somewhere else. Where that is, I'm not 100% sure, I'd assumed it was something that was installed with @supabase/supabase-js packaging, but I have no way to confirm that, or at least don't know of a way to confirm that.. 

I really am stumped by this, I have created plenty of realtime connections in the past and have never seen this one. I have deleted my entire node_modules folder and used yarn to re-install all with no changes.

**Reply by GaryAustin1** [reply]
So one clue from me is I'm using svelte-kit and I have whatwg-url (which is different than your similarly named module) as the node module.  node-fetch is the module who's package.json calls for whatwg-url.
I'm thinking something in your fetch environment is screening and not liking wss: or websockets.

**Reply by react-ghost** [reply]
Interesting. I'm not even sure how to go about fixing that, honestly. I can't think of one thing I've done differently in creating this project next to anything else I've created for react-native environments.. Unfortunately I have too much Javascript into the project to undo this. Deleting the backend and starting over was no big deal (for now) but the front end - not so much..

**Reply by react-ghost** [reply]
I found one thing. If I delete  `whatwg-url-without-unicode` it throws an error from a package that is required to use supabase with react native expo projects, which is `react-native-url-polyfill`.

This library is to be imported at the top of App.js/tsx, again, odd that it's throwing an error because i have other expo projects set up this way with no issues, but that library is causing issues, and yet, I can't use Supabase with react-native expo without that library.

**Reply by GaryAustin1** [reply]
Could be a new version of it?  Can you see the version in your working and non-working situations?

**Reply by react-ghost** [reply]
Yes, and they're the same for both the url-polyfill and I can make them the same with @supabase/supabase-js but it still makes no difference. Upgrading and downgrading version types all over and still no cigar.

**Reply by react-ghost** [reply]
Further updating this thread -


In Supabase's RealtimeClient.js there is a method called connect() which does a range of things I don't understand but in particular makes a call to to URL(url) which is a method provided by something in the  whatwg-url-api universe and that is exactly where the error is, as the URL being passed is null. The code from this section of Supabase's RealtimeClient.js is:

```
 connect() {
        var _a;
        if (this.conn) {
            return;
        }
        this.conn = new this.transport(this.endPointURL(), [], null, this.headers);
        if (this.conn) {
            // this.conn.timeout = this.longpollerTimeout // TYPE ERROR
            this.conn.binaryType = 'arraybuffer';
            this.conn.onopen = () => this._onConnOpen();
            this.conn.onerror = (error) => this._onConnError(error);
            this.conn.onmessage = (event) => this.onConnMessage(event);
            this.conn.onclose = (event) => this._onConnClose(event);
            // These are the problem lines
            const versionDate = new Date((_a = new URL(this.conn.url).searchParams.get('vsndate')) !== null && _a !== void 0 ? _a : '');
            if (versionDate instanceof Date) {
                this.versionDate = versionDate;
            }
        }
    }
```


It's all in relation to this version date. Am I really understanding that my progress on this project is being stonewalled because it's trying to extrapolate the version from the URL and that's it? When I comment the problem lines out the app works now and allows a realtime connection to be established.


Edit: I can confirm that this code from `RealtimeClient.js` does not exist in my working versions of supabase with React-Native.

```

const versionDate = new Date((_a = new URL(this.conn.url).searchParams.get('vsndate')) !== null && _a !== void 0 ? _a : '');
            if (versionDate instanceof Date) {
                this.versionDate = versionDate;
            }
```

**Reply by GaryAustin1** [reply]
That is definitely a change in realtime.js as of 10 days ago.
I verified that the new code is in my sveltekit app and no issues there.  
So some interaction between that new code and the different whatwg-url used in ReactNative.

**Reply by w3b6x9** [reply]
I’ll take a look at this and put in a fix today.

**Reply by react-ghost** [reply]
Interesting, it's my first time dealing with a bug that I didn't create myself so I was flabbergasted! What a wild ride. Thanks for sticking it out with me guys.

**Reply by w3b6x9** [reply]
@GaryAustin1 thanks for all your help with the debugging on this one!

**Reply by w3b6x9** [reply]
@adrienclay36 just deployed a fix to supabase-js. give the latest version a try.

**Reply by react-ghost** [reply]
@w3b6x9 Just updated to @supabase/supabase-js v 1.35.2 which looks like it's using realtime-js@1.7.2 and I can confirm that I am no longer receiving the error - It would've happened immediately at app load. Looks like you ended up maybe just removing those lines??

Looks like it's resolved though, thanks a million for that!

**Comment by GaryAustin1** [comment]
Yeah, I'm not trying to say change, just trying to narrow down the issue.  That module is not part of Supabase, so something about react native, versions of node, etc., but without just random googling on wss react native, react native supabase realtime.  Don't have more for you.

**Reply by react-ghost** [reply]
Okay, well that's a good start - thanks for that info. I would've certainly assumed it WAS apart of Supabase and didn't really know how to find out otherwise. Thanks a bunch for that.
